hilti/src/rt/tests/library.cc

```diff
@@ -0,0 +1,225 @@
+        REQUIRE_NE(path, nullptr);
+        _path = path;
+    }
+
+    ~TemporaryDirectory() {
+        if ( ! std::filesystem::exists(_path) )
+            return;
+
+        // Make sure we have permissions to remove the directory.
+        std::filesystem::permissions(_path, std::filesystem::perms::all);
+        for ( const auto& entry : std::filesystem::recursive_directory_iterator(_path) )
+            std::filesystem::permissions(entry, std::filesystem::perms::all);
+
+        std::filesystem::remove_all(_path);
+    }
+
+    const auto& path() const { return _path; }
+
+private:
+    std::filesystem::path _path;
+};
+
+// RAII helper to redirect output.
+class CaptureIO {
+public:
+    CaptureIO(std::ostream& stream) : _old(stream.rdbuf(_buffer.rdbuf())), _stream(&stream) {}
+    ~CaptureIO() { _stream->rdbuf(_old); }
+
+    auto str() const { return _buffer.str(); }
+
+private:
+    std::stringstream _buffer = std::stringstream{};
+    std::streambuf* _old = nullptr;
+    std::ostream* _stream = nullptr;
+};
+
+TEST_SUITE_BEGIN("Library");
+
+TEST_CASE("construct") {
+    Library _(dummy); // Does not throw
+    CHECK_THROWS_WITH_AS(Library("/does/not/exist"), "no such library: \"/does/not/exist\"", const EnvironmentError&);
+
+    SUBCASE("TMPDIR does not exist") {
+        Env _("TMPDIR", "/does/not/exist");
+        try {
+            Library _(dummy);
+            FAIL("expected exception not thrown");
+        } catch ( const EnvironmentError& e ) {
+            CAPTURE(e.description());
+            CHECK(startsWith(e.description(),
+                             fmt("could not add library \"%s\": could not create temporary file: ", dummy.c_str())));
+        } catch ( ... ) {
+            FAIL("unexpected exception thrown");
+        }
+    }
+
+    SUBCASE("cannot write TMPDIR") {
+        // NOTE: This test likely does not work if run as `root` as `root`
+        // can probably create files even in read-only directories.
+        TemporaryDirectory tmpdir;
+        std::filesystem::permissions(tmpdir.path(), std::filesystem::perms::none);
+        Env _("TMPDIR", tmpdir.path().c_str());
+
+        try {
+            Library _(dummy);
+            FAIL("expected exception not thrown");
+        } catch ( const EnvironmentError& e ) {
+            CAPTURE(e.description());
+            CHECK(startsWith(e.description(),
+                             fmt("could not add library \"%s\": could not create temporary file in ", dummy.c_str())));
+        } catch ( ... ) {
+            FAIL("unexpected exception thrown");
+        }
+    }
+}
+
+TEST_CASE("destruct") {
+    TemporaryDirectory tmp;
+    Env _("TMPDIR", tmp.path().c_str());
+
+    auto library = std::unique_ptr<Library>(new Library(dummy));
+
+    // Recursively change the permissions of the directory
+    // so the temporary library cannot be removed anymore.
+    //
+    // NOTE: This test likely does not work if run as `root` as `root`
```

Comment:
same. 

Does doctest have a notion of "run test only of X is true"?

Reply:
See above, dropping.

No, I don't think doctest even has a way to explicitly `SKIP` tests. One could probably add a `return` or `continue` statement to skip the rest of a test (probably after emitting a `WARNING`).

What we could do if this becomes a repeated issue is to update the test filter by providing a custom test [main function](https://github.com/onqtam/doctest/blob/master/doc/markdown/main.md).